lib/hood/kiln.hoon

```diff
@@ -388,33 +381,39 @@
     ==  ==
   ::
   ++  start-sync
-    |=  reset/?
-    =.  +>.$  (spam (render "activated sync" sud her syd) ~)
-    %-  blab
-    :~  :*  ost  %warp
-            [%kiln %sync syd (scot %p her) sud ?:(reset /reset /)]
-            [our her]  sud  ~  %sing  %w  [%da now]  /
-    ==  ==
+    =<  (spam (render "activated sync" sud her syd) ~)
+    =/  =wire  /kiln/sync/[syd]/(scot %p her)/[sud]
+    (blab [ost %warp wire [our her] sud `[%sing %w [%da now] /]] ~)
   ::
   ++  writ
-    |=  {reset/? rot/riot}
+    |=  rot=riot
     ?~  rot
       %^    spam
           leaf+"bad %writ response"
         (render "on sync" sud her syd)
       ~
     =.  let  ?.  ?=($w p.p.u.rot)  let  ud:((hard cass:clay) q.q.r.u.rot)
-    %-  blab  ^-  (list move)  :_  ~
-    :*  ost  %merg
-        [%kiln %sync syd (scot %p her) sud ?:(reset /reset /)]
-        our  syd  her  sud  ud+let
-        ?:  =(0 ud:.^(cass:clay %cw /(scot %p our)/[syd]/(scot %da now)))
-          %init
-        %mate
-    ==
+    =/  =wire  /kiln/sync/[syd]/(scot %p her)/[sud]
+    =/  =cass  .^(cass:clay %cw /(scot %p our)/[syd]/(scot %da now))
+    ::
+    ::    If we will be syncing in remote changes, we need all our sync merges
```

Comment:
This is pretty gross, and I don't completely get it. It seems like it might be Clay's problem more than a boot problem, though.

Reply:
This is indeed gross, and the heart of the problem. Clay doesn't know about syncs, they're a userspace feature implemented here. Maybe clay should have a merge mode that starts normally (as a `%mate` or whatever), but escalates all the way up to %init. It'd still have to be triggered from somewhere, either here in kiln or in the init logic of %dill.

Similarly, I don't know how we would block user edits until a sync succeeds; %clay doesn't know about syncs. It wouldn't be enough even to block user edits until a merge has succeeded once on every desk for all non-galaxy ships -- %home and %kids are created through merges, and receive the first two syncs from the pill or arvo directory as merges from %base. It seems like we're forced to hardcode a threshold like this, unless/until we have a smarter merge type that can synthesize a merge-base.

Reply:
Ok, that makes a little more sense. It sounds like we'll be more or less stuck with this until we move sync from userspace into Clay itself, which seems more and more like the thing to do. I'm willing to accept the necessity of this hack for now.

What is it about receiving the userspace files from the pill instead of from the local Arvo that forces us to do this weird hardcoded logic? Current release-candidate does something during boot to get Clay up to speed that doesn't require this.

Reply:
I've read this a few more times and have a better sense of it now. It's really brittle, but looks like it should work fine unless the user does make a commit before the sync happens. Ensuring the order of Clay transactions is a problem we'll have to solve more generally at some point anyway.